File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxlen = 0;
        int hash[]=new int[128];
         Arrays.fill(hash, -1);
        int l = 0,r=0;
        while(r<s.length())
        {
            char ch=s.charAt(r);
            if(hash[ch]!=-1)
            {
                if(hash[ch]>=l)
                l=hash[ch]+1;
            }
            hash[ch]=r; //store the index
            maxlen=Math.max(maxlen,r-l+1);
            r++;
        }
       return maxlen;
        }
     
        
 
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size(), i = 0, j = 0, maxi = 0, k;
        map<char, int> mpp;
        // if(s=='')
        // return 0;
        while (j < n) {
            if (mpp.find(s[j]) != mpp.end()) {
                k = mpp[s[j]];
                if(i<=k)
                { 
                 i=k+1;
                }
            }
            mpp[s[j]]=j;
            maxi = max(maxi, j - i + 1);
            j++;
        }
        
        return maxi;
    }
};